roles/translator/src/main.rs

```diff
@@ -54,22 +54,122 @@ async fn main() {
-    let (tx_sv1_bridge, rx_sv1_downstream) = unbounded();
+    let target = Arc::new(Mutex::new(vec![0; 32]));
+
+    // Sender/Receiver to send SV1 `mining.notify` message from the `Bridge` to the `Downstream`
+    let (tx_sv1_notify, _rx_sv1_notify): (
+        broadcast::Sender<server_to_client::Notify>,
+        broadcast::Receiver<server_to_client::Notify>,
+    ) = broadcast::channel(10);
+
+    let task_collector: Arc<Mutex<Vec<(AbortHandle, String)>>> = Arc::new(Mutex::new(Vec::new()));
 
+    start(
+        tx_sv1_notify.clone(),
+        target.clone(),
+        tx_status.clone(),
+        task_collector.clone(),
+        proxy_config.clone(),
+    )
+    .await;
+
+    debug!("Starting up signal listener");
+    let task_collector_ = task_collector.clone();
+
+    let mut interrupt_signal_future = Box::pin(tokio::signal::ctrl_c().fuse());
+    debug!("Starting up status listener");
+    // Check all tasks if is_finished() is true, if so exit
+    loop {
+        let task_status = select! {
+            task_status = rx_status.recv().fuse() => task_status,
+            interrupt_signal = interrupt_signal_future => {
+                match interrupt_signal {
+                    Ok(()) => {
+                        info!("Interrupt received");
+                    },
+                    Err(err) => {
+                        error!("Unable to listen for interrupt signal: {}", err);
+                        // we also shut down in case of error
+                    },
+                }
+                break;
+            }
+        };
+        let task_status: Status = task_status.unwrap();
+
+        match task_status.state {
+            // Should only be sent by the downstream listener
+            State::DownstreamShutdown(err) => {
+                error!("SHUTDOWN from: {}", err);
+                break;
+            }
+            State::BridgeShutdown(err) => {
+                error!("SHUTDOWN from: {}", err);
+                break;
+            }
+            State::UpstreamShutdown(err) => {
+                error!("SHUTDOWN from: {}", err);
+                break;
+            }
+            State::UpstreamTryReconnect(err) => {
+                error!("SHUTDOWN from: {}", err);
+
+                // wait a random amount of time between 0 and 3000ms
+                // if all the downstreams try to reconnect at the same time, the upstream may fail
+                tokio::time::sleep(Duration::from_millis(1000)).await;
+                let mut rng = rand::thread_rng();
+                let wait_time = rng.gen_range(0..=3000);
+                tokio::time::sleep(Duration::from_millis(wait_time)).await;
+
+                // kill al the tasks
+                let task_collector_aborting = task_collector_.clone();
+                kill_tasks(task_collector_aborting.clone());
+
+                error!("Trying recconnecting to upstream");
```

Comment:
Shouldn't be a warning?
Also, there's a typo: it should be "Trying reconnecting to upstream"